activesupport/lib/active_support/core_ext/erb/util.rb

```diff
@@ -124,7 +122,12 @@ def html_escape_once(s)
     # JSON gem, do not provide this kind of protection by default; also some gems
     # might override +to_json+ to bypass Active Support's encoder).
     def json_escape(s)
-      result = s.to_s.gsub(JSON_ESCAPE_REGEXP, JSON_ESCAPE)
+      result = s.to_s.dup
+      result.gsub!(">", '\u003e')
+      result.gsub!("<", '\u003c')
+      result.gsub!("&", '\u0026')
+      result.gsub!("\u2028", '\u2028')
+      result.gsub!("\u2029", '\u2029')
```

Comment:
This is very interesting, especially since in the past merging regexps did bring improvements, but I guess here being able to call `gsub` with a String instead makes all the difference. I might dig in `gsub!` later to double check why.

Reply:
I think it depends case-by-case. In this case I think we're avoiding significant allocations in the gsub when there are replacements to make (2 per match), but this held up even on strings which didn't need escaping, which I think is due to the unicode ().

I asked @HParker who pointed out that a different instruction is used within the regexp engine (`OP_CCLASS_MIX` ?) when the character class being matched includes non-ASCII.

Reply:
>  I think we're avoiding significant allocations in the gsub when there are replacements to make 

So somewhat orthogonal but I think it would be possible to avoid these allocations in gsub in such case by re-using the `setup_fake_str` code used for looking up interned strings. Could be a fun patch to work on.

Reply:
I agree! One complexity is that the map may have a default proc which would allow the value to escape (we have to be similarly careful with the `opt_aref_with` insn)